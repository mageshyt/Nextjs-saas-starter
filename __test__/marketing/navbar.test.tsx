import { render, screen } from '@testing-library/react';
import { describe, it, expect, vi } from 'vitest';
import { useUser } from '@clerk/nextjs';
import Navbar from '@/app/(marketing)/components/navbar';

vi.mock('@clerk/nextjs', async () => {
  const actual = await vi.importActual('@clerk/nextjs');
  return {
    ...actual,
    useUser: vi.fn(),
    UserButton: () => <button>Mocked UserButton</button>, // Mock UserButton
  };
});
describe('Navbar', () => {

  it('renders the logo image with the correct alt text', () => {
    (useUser as ReturnType<typeof vi.fn>).mockReturnValue({ isSignedIn: false });
    render(<Navbar />);
    const logo = screen.getByAltText('logo');
    expect(logo).toBeInTheDocument();
  });

  it('shows Sign In and Sign Up buttons if the user is signed out', () => {
    (useUser as ReturnType<typeof vi.fn>).mockReturnValue({ isSignedIn: false });
    render(<Navbar />);
    expect(screen.getByText('Sign In')).toBeInTheDocument();
    expect(screen.getByText('Sign up')).toBeInTheDocument();
  });

  it('renders UserButton when the user is signed in', () => {
    (useUser as ReturnType<typeof vi.fn>).mockReturnValue({ isSignedIn: true });
    render(<Navbar />);
    expect(screen.getByText('Mocked UserButton')).toBeInTheDocument();
  });

  it("renders 'Dashboard' link when user is signed in", () => {
    (useUser as ReturnType<typeof vi.fn>).mockReturnValue({ isSignedIn: true });
    render(<Navbar />);
    const dashboardLink = screen.getByRole('link', { name: /dashboard/i });   
     expect(dashboardLink).toBeInTheDocument();
    expect(dashboardLink).toHaveAttribute('href', '/dashboard'); // Optional
  });

  it('handles uninitialized user state gracefully', () => {
    (useUser as ReturnType<typeof vi.fn>).mockReturnValue(null);
    render(<Navbar />);
    expect(screen.queryByText('Sign In')).toBeInTheDocument();
  });

  it('matches the snapshot when user is signed in', () => {
    (useUser as ReturnType<typeof vi.fn>).mockReturnValue({ isSignedIn: true });
    const { container } = render(<Navbar />);
    expect(container).toMatchSnapshot();
  });
});