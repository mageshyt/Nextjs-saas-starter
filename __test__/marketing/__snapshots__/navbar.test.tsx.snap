// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Navbar > matches the snapshot when user is signed in 1`] = `
<div>
  <header
    class="fixed top-0 left-0 z-50 w-full animate-fade-in opacity-0 border-b p-2  backdrop-blur-[12px] [--animation-delay:600ms]"
  >
    <div
      class="flex h-[3rem] mx-4  items-center justify-between"
    >
      <a
        href="/"
      >
        <div
          class="relative size-6"
        >
          <img
            alt="logo"
            class="object-cover"
            data-nimg="fill"
            decoding="async"
            loading="lazy"
            src="/logo.svg"
            style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
          />
        </div>
      </a>
      <div
        class="ml-auto gap-6 flex h-full items-center"
      >
        <a
          class="relative inline-flex h-10 overflow-hidden rounded-full p-[2px] focus:outline-none  "
          href="/dashboard"
        >
          <span
            class="absolute inset-[-1000%] animate-[spin_2s_linear_infinite] bg-[conic-gradient(from_90deg_at_50%_50%,#E2CBFF_0%,#393BB2_50%,#E2CBFF_100%)]"
          />
          <span
            class="inline-flex h-full w-full cursor-pointer items-center justify-center rounded-full bg-slate-950 px-3 py-1 text-sm font-medium text-white backdrop-blur-3xl"
          >
            Dashboard
          </span>
        </a>
        <button>
          Mocked UserButton
        </button>
      </div>
    </div>
  </header>
</div>
`;
